File: src/main/java/fr/lacombe/ItemReference.java
package fr.lacombe;

class ItemReference {

    private final String itemCode;
    private final Price itemPrice;

    private ItemReference(String itemCode, Price itemPrice) {
        this.itemCode = itemCode;
        this.itemPrice = itemPrice;
    }

    static Builder aBuilder() {
        return new Builder();
    }

    String getItemCode() {
        return itemCode;
    }

    Price getItemPrice() {
        return itemPrice;
    }

    static final class Builder {
        private String itemCode;
        private Price itemPrice;

        private Builder() {
        }

        Builder withItemCode(String itemCode) {
            this.itemCode = itemCode;
            return this;
        }

        Builder withItemPrice(double itemPrice) {
            this.itemPrice = Price.valueOf(itemPrice);
            return this;
        }

        ItemReference build() {
            return new ItemReference(itemCode, itemPrice);
        }
    }
}

File: CaisseEnregistreuse/CaisseEnregistreuse/ItemReference.cs
using System.Collections.Immutable;

namespace CaisseEnregistreuse
{
    public class ItemReference
    {
        public readonly string Name;
        public readonly Price Price;

        public ItemReference(string name, double price)
        {
            Name = name;
            Price = Price.ValueOf(price);
        }

        private ItemReference(string itemCode, Price unitPrice)
        {
            Name = itemCode;
            Price = unitPrice;
        }

        public bool MatchByItemCode(string name)
        {
            return Name == name;
        }

        public static ItemReferenceBuilder Builder()
        {
            return new ItemReferenceBuilder();
        }

        public sealed class ItemReferenceBuilder
        {
            private string _itemCode;
            private Price _unitPrice;

            internal ItemReferenceBuilder() {}

            public ItemReferenceBuilder WithItemCode(string itemCode)
            {
                this._itemCode = itemCode;
                return this;
            }

            public ItemReferenceBuilder WithUnitPrice(double unitPrice)
            {
                return WithUnitPrice(Price.ValueOf(unitPrice));
            }

            ItemReferenceBuilder WithUnitPrice(Price unitPrice)
            {
                this._unitPrice = unitPrice;
                return this;
            }

            public ItemReference Build()
            {
                return new ItemReference(_itemCode, _unitPrice);
            }
        }
    }
}